### PR TITLE
Ce/loc reassignment ui

### PR DESCRIPTION
**Pull request by calellowitz**
This represents the UI work for dashboard location reassignment

As such, im basing this PR off https://github.com/dimagi/commcare-hq/pull/27083, rather than master so that it includes the relevant data model changes.
##### SUMMARY

The PR generally has 4 parts, each of which is intended to be contained to a single commit. 

There are some changes to the django views that the dashboard front end uses to get location information. The views now also return information about the locations that any returned locations replace or are replaced by, including the date of the deprecation and the full hierarchy names of the other locations.

This information is then used on the front end. I created a new directive that templates a message for users, which will be the same in all places for now. The language has yet to be finalized and there will likely be two messages depending on whether the location is replacing another or has been replaced. 

This message is displayed whenever someone selects a location that is not active instead of the data that would otherwise show in the report, as well as in the location filter. This part is pretty repetitive, and I tried to use shared functionality when I could, but there is little to nothing shared between all these directives. I added it to the baseFilterController, which is used in most of them, but had to individually add the logic to those that did not use that. I am definitely open to suggestions for how to DRY up the javascript.

The final change is a small change to the fact sheets for when a location is inactive for part of the three months displayed, but not all of them.

This can be reviewed 🐡 

The general spec for location reassignment work is https://docs.google.com/document/d/1h6kTr9M0FumswP3-gr_84OLeUHRmn5Y10dCpSAS3Y9g/edit#

Dashboard specific information is also in https://docs.google.com/document/d/1LO9oO3Z-JQbeTMr-76sgyVDMFfYWM5eItEvTxJq2E2Y/edit

### CODE REVIEW

File: custom/icds_reports/static/js/angular-services/baseControllers.service.js
Comment (mkangia) [inline custom/icds_reports/static/js/angular-services/baseControllers.service.js:65]: Thinking from how I imagined this to be visible on dashboard, should not there be two different methods (showDeprecatedMessage & showDeprecatesMessage) to show message for a 
1. deprecated location
2. a location that deprecated another location

I was thinking of two different messages like

on a deprecated location we see
```
this location was extracted from ABC or
this location was split from ABC or
this location was a result of merge between ABC and DEF or
this location was moved from ABC
```

and

on the other location we see similar or simple like
```
this location deprecated DEF via merge/split/extract/move
```

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (mkangia) [inline custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:2]: I think we should not use a new term like `replaced` and rather use merge/split/extract/move or just `deprecate`. Plus replaced would be confusing for some operation which does not really replace a location but get extracted from it like in A -> A + B

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1794]: nit: it would be nice to put these numbers in constants. I never remember if 1 is state or 1 is AWC but a constant name would be useful.

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1788]: nit : `repacement`

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1811]: is it better to have inline functions declared on top?

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1819]: nit: `repacement`

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1821]: `parent__parent__parent__parent` 🔢

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [inline custom/icds_reports/utils/__init__.py:1856]: how about `get_location_names_from_metafield`?

**Review by mkangia — COMMENTED** [review]
reviewed mainly from feature and code perspective and not from how this fits in dashboard since i am not well aware of that.
Overall i think we should avoid using `replace`, introducing a new term to vocabulary we have been using, which we will lead to confusion. `deprecate` for locations and move/split/extract/merge as operations.

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1821]: Could be 4 parents. better safe than sorry

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1811]: I declared it here because it was used once and this is right before that use. I could be convinced it should be higher up

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1856]: I agree this name can be improved, but the replacement names arent actually coming from the meta, they are coming from the replacment_names dict I pass in, so I don't love that option. I will think a bit on this

File: custom/icds_reports/static/js/angular-services/baseControllers.service.js
Comment (calellowitz) [reply custom/icds_reports/static/js/angular-services/baseControllers.service.js:65]: I think I am going to leave most of that logic for the angular template. `reassignment-message.html` can figure out how to render the message since it has the selected location, and this keeps the logic in the main controllers and the rest of the templates much simpler.

**Review by czue — COMMENTED** [review]
Looks good overall, just had mostly minor stuff. The duplication on the front-end is sad, but I agree that there's not an easy path to fixing that.

How hard would it be to add some tests for the utility functions that pull deprecation info? I think those are the hairiest parts of this code and tests would at least serve as a form of documentation if the code itself can't be made more clear.

File: custom/icds_reports/views.py
Comment (calellowitz) [reply custom/icds_reports/views.py:700]: whatever was going on in https://github.com/dimagi/commcare-hq/pull/27166/files/bcb787e684dac9d7434dccacc064d97fb98c2cdb#diff-129264561592e648289c087ab1350b9bL699 (I just pulled it out of the dict comprehension), but it is combining into a single list the set of locations that will be returned by this function. I think that is basically the selected location, all its sibling locations, its parents going back to the top level, plus all their siblings. Basically, everything needed to seed the locations in the location filter on initial page load, which I think is the only time this is called.

File: custom/icds_reports/templates/icds_reports/dashboard.html
Comment (calellowitz) [reply custom/icds_reports/templates/icds_reports/dashboard.html:92]: Is that instead of here or in addition to here?

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1856]: yeah

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1815]: This is just a simple map of `{location_id: location_name}`. What makes it different than just putting the name is that it includes the whole hierarchy of that location, so that users can find it in the location filter. It looks like `'state > district > block'`. I can add that in a comment, if you think that would help since its obviously not clear thats what it does, but an object seems like overkill

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1794]: done

**Comment by calellowitz** [comment]
@mkangia @czue  I think this is also ready for review again, although the message text is still not totally finalized

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [reply custom/icds_reports/utils/__init__.py:1801]: Seeing other instances on HQ i see they are declared in the beginning. Does that convince you?

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [reply custom/icds_reports/utils/__init__.py:1846]: Just considering this method as a standalone where it does not know the intention of the names so if we rename method name and argument as
```
def get_location_names_from_metafield(location, field, location_names)
```
it will sound better.

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (mkangia) [inline custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:1]: aha! this is better.

nit: i think we should check on `deprecated_by` and `deprecates` even though its true that if those are set, the timestamp fields would be too.

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (mkangia) [reply custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html]: @calellowitz have you considered this yet? I did comment this on the html but the idea was overall even in code.

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [reply custom/icds_reports/utils/__init__.py]: @calellowitz did you think about this?

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (mkangia) [reply custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:1]: I am surprised, we aren't using the operation available on the location?
I see value in telling the user what operation was performed like
`It was merged extracted from ABC on 1-1-2001`

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (mkangia) [inline custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:5]: just confirming that here `deprecates` and `deprecated_to` are location names and not the id? I guess you already mapped that somewhere but could not point to it, so just confirming.

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1801]: good enough for me

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py]: oops, that idea sounds good, and we actually already do have constants for those values. will add them

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (calellowitz) [reply custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html]: Any conversation about language in these messages probably needs to be resolved on https://docs.google.com/document/d/1LO9oO3Z-JQbeTMr-76sgyVDMFfYWM5eItEvTxJq2E2Y/edit with Nisha and Nitish. That language is mostly their language, but I will flag that they are explicitly not in favor of the word `deprecate` which is why it was removed in the last commit.

File: custom/icds_reports/utils/__init__.py
Comment (czue) [reply custom/icds_reports/utils/__init__.py]: sorry, that seems fine, I got confused by [this line](https://github.com/dimagi/commcare-hq/pull/27166/files#diff-c87f5ffa4bc46d5bec44951bdde6be41R1847) and thought it had some additional structure, but see now that's coming from `location.metadata`

File: custom/icds_reports/templates/icds_reports/dashboard.html
Comment (czue) [reply custom/icds_reports/templates/icds_reports/dashboard.html:92]: instead of

File: custom/icds_reports/views.py
Comment (czue) [reply custom/icds_reports/views.py:676]: Sorry, I guess a more specific question is, is there a reason it's not just:

`all_locations = list(locations) + list(parents)`

**Comment by czue** [comment]
@calellowitz did you have any thoughts on adding some tests?

File: custom/icds_reports/views.py
Comment (calellowitz) [reply custom/icds_reports/views.py:676]: It does seem like that will work, I am not super confident in whatever transformations are happening here, but it appears that is equivalent. And certainly simpler to read. Maybe it enforced ordering that wasn't guaranteed in python 2

File: custom/icds_reports/templates/icds_reports/dashboard.html
Comment (calellowitz) [reply custom/icds_reports/templates/icds_reports/dashboard.html:92]: done

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (calellowitz) [reply custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:1]: > nit: i think we should check on deprecated_by and deprecates even though its true that if those are set, the timestamp fields would be too.

I did it this way, in case the field of what it was replaced with is not properly set. The date determines whether or not it was deprecated. If we can't figure out what it was deprecated to (somehow the location got deleted or something else) it should not prevent us from letting the user know they are viewing an invalid location.

File: custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html
Comment (calellowitz) [reply custom/icds_reports/templates/icds_reports/icds_app/reassignment-message.html:5]: Yeah, thats what the util changes do (the function `get_deprecation_info` is for finding the names of the relevant locations)

**Comment by calellowitz** [comment]
> @calellowitz did you have any thoughts on adding some tests?

Adding tests sounds like a great idea, although not one I currently know how to do. I don't want to block QA on the time it takes me to figure that out, but happy to add some in a follow up, because this definitely feels like some tests could help

File: custom/icds_reports/views.py
Comment (czue) [reply custom/icds_reports/views.py:676]: Ok. My personal opinion is might as well get to the bottom of it and clear it up now so as not to confuse future people looking at this, but if that feels too difficult then I don't think it needs to block.

**Comment by czue** [comment]
> I don't want to block QA on the time it takes me to figure that out, but happy to add some in a follow up, because this definitely feels like some tests could help

This sounds fine to me, definitely doesn't need to block QA

**Review by czue — COMMENTED** [review]
I'm signed off on this - at least to the extent that I feel there are strong diminishing returns on time with each additional review pass I'm doing.

That said, I'd like to see an approval from @mkangia too (which is why I'm not approving) since I think his context is important here.

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1846]: How about `get_location_replacement_name` that makes it more clear that we are talking about locations, while not implying the name is in the meta field?

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py:1801]: moved

File: custom/icds_reports/utils/__init__.py
Comment (calellowitz) [reply custom/icds_reports/utils/__init__.py]: added

**Comment by calellowitz** [comment]
@mkangia I think I got to all the comments. Let me know what you think. Looking to get this out this week for QA

File: custom/icds_reports/utils/__init__.py
Comment (mkangia) [reply custom/icds_reports/utils/__init__.py]: That earlier name suggestion came from combining the two things
1. avoid using replacement
2. make method name descriptive of what its doing

> but the replacement names aren't actually coming from the meta

Yup, but the location id is in the metafield, so I included that in the method name.

`get_location_replacement_name` partially conveys what the method is doing which would be confusing, it is getting location name from location_names but how does it pick it? but its better than `get_replacement_name`
(this is non blocking)

**Review by mkangia — APPROVED** [review]
Defer to @czue from dashboard implementation perspective, looks good from the comments i had

**Comment by calellowitz** [comment]
Since this is approved, going to merge it into the other PR so tests run on the full thing. Will hold off merging that one until for now, but expect to merge it soon

**Pull request merged**
